src/Core/Framework/App/Context/Gateway/AppContextGateway.php

```
@@ -0,0 +1,123 @@
+    {
+        if (!$payload->getData()->get('appName')) {
+            throw AppException::missingRequestParameter('appName');
+        }
+
+        $appName = $payload->getData()->get('appName');
+        $app = $this->getApp($appName, $payload->getContext()->getContext());
+
+        $appPayload = new AppContextGatewayPayload($payload->getContext(), $payload->getCart(), $payload->getData()->all());
+
+        /** @var string $checkoutGatewayUrl */
+        $checkoutGatewayUrl = $app->getContextGatewayUrl();
+        $appResponse = $this->payloadService->request($checkoutGatewayUrl, $appPayload, $app);
+
+        if (!$appResponse) {
+            throw AppException::gatewayRequestFailed($app->getName(), 'context');
+        }
+
+        $commands = $this->collectCommandsFromAppResponse($appResponse);
+
+        return $this->executor->execute($commands, $payload->getContext());
+    }
+
+    private function getApp(string $appName, Context $context): AppEntity
+    {
+        $criteria = new Criteria();
+        $criteria->addFilter(new EqualsFilter('name', $appName));
+        $criteria->addFilter(new EqualsFilter('active', true));
+
+        /** @var AppEntity|null $app */
+        $app = $this->appRepository->search($criteria, $context)->first();
+
+        if (!$app) {
+            throw AppException::appNotFoundByName($appName);
+        }
+
+        if (!$app->getContextGatewayUrl()) {
+            throw AppException::gatewayNotConfigured($appName, 'context');
+        }
+
+        return $app;
+    }
+
+    private function collectCommandsFromAppResponse(AppContextGatewayResponse $response): ContextGatewayCommandCollection
+    {
+        $collected = new ContextGatewayCommandCollection();
+
+        foreach ($response->getCommands() as $payload) {
+            if (!isset($payload['command'], $payload['payload'])) {
+                $this->logger->logOrThrowException(GatewayException::payloadInvalid($payload['command'] ?? null));
```

I know this is probably being pedantic, but could you add negative tests too and validate that we are logging when you expect it to log?

---

@Bird87ZA I am. By expecting exceptions, I already tested, that the ExceptionLogger is indeed running. In Test environments it will throw exceptions, rather than only logging it. 

The logger in itself is properly tested, therefore I see no further need here